go-controller/hybrid-overlay/pkg/controller/node_linux.go

```diff
@@ -0,0 +1,477 @@
+	}
+
+	_, _, err = util.RunOVSOfctl("add-flow", extBridgeName,
+		fmt.Sprintf("cookie=0x%s,table=0,priority=100,ip,nw_dst=%s,actions=load:4097->NXM_NX_TUN_ID[0..31],set_field:%s->tun_dst,set_field:%s->eth_dst,output:1", cookie, cidr.String(), nodeIP.String(), drMAC.String()))
+	if err != nil {
+		return fmt.Errorf("failed to add VXLAN flow for node %q: %v", node.Name, err)
+	}
+
+	return nil
+}
+
+// Add handles node additions
+func (n *NodeController) Add(node *kapi.Node) {
+	if node.Name == n.nodeName {
+		if err := n.ensureHybridOverlayBridge(); err != nil {
+			logrus.Errorf(err.Error())
+		}
+	} else {
+		if err := n.windowsNodeAddOrUpdate(node); err != nil {
+			logrus.Warning(err)
+		}
+	}
+}
+
+// Update handles node updates
+func (n *NodeController) Update(oldNode, newNode *kapi.Node) {
+	if newNode.Name != n.nodeName {
+		if nodeChanged(oldNode, newNode) {
+			if err := n.windowsNodeAddOrUpdate(newNode); err != nil {
+				logrus.Warning(err)
+			}
+		}
+	}
+}
+
+func deleteNodeFlowsByCookie(cookie string) error {
+	_, _, err := util.RunOVSOfctl("del-flows", extBridgeName, fmt.Sprintf("table=0, cookie=0x%s/0xffffffff", cookie))
+	if err != nil {
+		return fmt.Errorf("failed to delete flows for node cookie %q: %v", cookie, err)
+	}
+	return nil
+}
+
+// Delete handles node deletions
+func (n *NodeController) Delete(node *kapi.Node) {
+	if node.Name == n.nodeName || !houtil.IsWindowsNode(node) {
+		return
+	}
+
+	if err := deleteNodeFlowsByCookie(nodeNameToCookie(node.Name)); err != nil {
+		logrus.Errorf(err.Error())
+	}
+}
+
+// Sync handles local node initialization and removing stale nodes on startup
+func (n *NodeController) Sync(nodes []*kapi.Node) {
+	// First ensure our local hybrid overlay is initialized
+	for _, node := range nodes {
+		if node.Name == n.nodeName {
+			if err := n.ensureHybridOverlayBridge(); err != nil {
+				logrus.Errorf(err.Error())
+			}
+			break
+		}
+	}
+
+	kubeNodes := make(map[string]bool)
+	for _, node := range nodes {
+		if houtil.IsWindowsNode(node) {
+			kubeNodes[nodeNameToCookie(node.Name)] = true
+		}
+	}
+
+	stdout, stderr, err := util.RunOVSOfctl("dump-flows", extBridgeName, "table=0")
+	if err != nil {
+		logrus.Errorf("failed to dump flows for %s: stderr: %q, error: %v", extBridgeName, stderr, err)
+		return
+	}
+
+	// Find all flows that exist in br-ext that are for nodes not present
+	// in the Kube node list
+	lines := strings.Split(stdout, "\n")
+	nodesToRemove := make(map[string]bool)
+	for _, line := range lines {
+		// Ignore the end-of-table drop rule
+		if strings.Contains(line, "actions=drop") {
+			continue
+		}
+
+		parts := strings.Split(line, ",")
+		for _, part := range parts {
+			part = strings.TrimSpace(part)
+			const cookieTag string = "cookie=0x"
+			if !strings.HasPrefix(part, cookieTag) {
+				continue
+			}
+			cookie := part[len(cookieTag):]
+			if len(cookie) != 8 {
+				// Ignore non-node-specific rules (eg cookie=0x0)
```

Comment:
it would be cleaner to put all the node-specific rules in their own table

Reply:
@danwinship yeah, can we do that in a separate PR?

Reply:
sure